### PR TITLE
Add unit test to current enforcement stats accounting

### PR DESCRIPTION
**Pull request by veshkemburu**
## Summary

Added three unit tests for enforcement stats accounting, to handle stats reporting for various cases.

1. On reporting rule stats for old versions
2. On reporting rule stats continuously while updating rule versions
3. Sending erroneous data(such as metrics lower than previously recorded)

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: lte/gateway/python/magma/pipelined/tests/test_enforcement_stats.py [hunk 1]
```diff
@@ -679,6 +680,55 @@ def test_rule_reactivation(self):
         #                 num_pkts_tx_match * len(packet))
         self.assertEqual(len(stats), 2)
 
+    def test_cookie_poll(self):
```
Comment (koolzz) [inline lte/gateway/python/magma/pipelined/tests/test_enforcement_stats.py]: Hm this was already commited, try rebasing.

File: lte/gateway/c/session_manager/test/test_local_enforcer.cpp [hunk 2]
```diff
@@ -443,13 +443,180 @@ TEST_F(LocalEnforcerTest, test_multi_version_reporting) {
...
+      create_update_tunnel_ids_request(IMSI1, BEARER_ID_1, teids1));
+
+  insert_static_rule(1, "", "rule1");
+  RuleRecordTable table;
+  auto record_list = table.mutable_records();
+  // ipv4 usage
+  create_rule_record(
+      IMSI1, test_cfg_.common_context.ue_ipv4(), "rule1", 1, 15, 30, 0, 0,
+      record_list->Add());
+  // ipv6 usage for the same charging key and subscriber
```
Comment (koolzz) [inline lte/gateway/c/session_manager/test/test_local_enforcer.cpp]: Lets just do ipv4 only for this one, we already do ipv6 in another test

**Review by koolzz — CHANGES_REQUESTED** [review]
looks good can you rebase though, and address linter

**Comment by themarwhal** [comment]
Can you run `make format_session_manager` in your VM? Should get rid of most of the cpp lints.

File: lte/gateway/c/session_manager/test/test_local_enforcer.cpp [hunk 3]
```diff
@@ -443,13 +443,180 @@ TEST_F(LocalEnforcerTest, test_multi_version_reporting) {
...
+  create_rule_record(
+      IMSI1, test_cfg_.common_context.ue_ipv6(), "rule2", 1, 15, 95, 0, 0,
+      record_list->Add());
+
+  auto update = SessionStore::get_default_session_update(session_map);
+  auto uc     = get_default_update_criteria();
+  //update to version 3 and check rx and tx values for older versions
+  session_map[IMSI1][0]->increment_rule_stats("rule1", &uc);
+  session_map[IMSI1][0]->increment_rule_stats("rule1", &uc);
+  session_map[IMSI1][0]->increment_rule_stats("rule1", &uc);
```
Comment (themarwhal) [inline lte/gateway/c/session_manager/test/test_local_enforcer.cpp:498]: Let's use lower case with underscore as outlined here: https://google.github.io/styleguide/cppguide.html#Variable_Names

File: lte/gateway/c/session_manager/test/test_local_enforcer.cpp [hunk 4]
```diff
@@ -443,13 +443,180 @@ TEST_F(LocalEnforcerTest, test_multi_version_reporting) {
...
-  // update to version2
+  // update to version3
   session_map[IMSI1][0]->increment_rule_stats("rule1", &uc);
   local_enforcer->aggregate_records(session_map, table, update);
   assert_charging_credit(session_map, IMSI1, SESSION_ID_1, USED_RX, {{1, 30}});
   assert_charging_credit(session_map, IMSI1, SESSION_ID_1, USED_TX, {{1, 140}});
+  EXPECT_EQ(session_map[IMSI1][0]->get_policy_stats("rule1").current_version, 3);
+}
+
+TEST_F(LocalEnforcerTest, test_old_version_reporting) {
```
Comment (themarwhal) [inline lte/gateway/c/session_manager/test/test_local_enforcer.cpp:455]: Can you add a summary of what the tests does?

**Comment by uri200** [comment]
Recommend to do
```
git pull upstream master
git push origin master
git checkout veshkemburu:unit_test_branch
git rebase master
git rebase -i master
# -i for interactive in case you need to squash
```

Also in many occasions it is better just to amend instead of creating a commit. So you can do this 
```
git commit --amend -s
```
That will overwrite your previous commit with the new changes, leaving your history clean

File: lte/gateway/c/session_manager/test/test_local_enforcer.cpp [hunk 5]
```diff
@@ -429,28 +429,216 @@ TEST_F(LocalEnforcerTest, test_multi_version_reporting) {
...
   local_enforcer->aggregate_records(session_map, table, update);
   assert_charging_credit(session_map, IMSI1, SESSION_ID_1, USED_RX, {{1, 30}});
   assert_charging_credit(session_map, IMSI1, SESSION_ID_1, USED_TX, {{1, 140}});
+  EXPECT_EQ(
+      session_map[IMSI1][0]->get_policy_stats("rule1").current_version, 3);
+}
+
+TEST_F(LocalEnforcerTest, test_old_version_reporting) {
+  /*
+  Tests reporting of tx and rx values of old versions, after
```
Comment (koolzz) [inline lte/gateway/c/session_manager/test/test_local_enforcer.cpp]: could you fix the style to this

```
/*
 * Tests reporting of tx and rx values of old versions, after
 * updating to a new version, as well as whether rule records of old
 * versions are disregarded appropriately
 */

File: lte/gateway/c/session_manager/test/test_local_enforcer.cpp [hunk 6]
```diff
@@ -429,28 +429,216 @@ TEST_F(LocalEnforcerTest, test_multi_version_reporting) {
...
+      session_map, IMSI1, SESSION_ID_1, test_cfg_, response);
+  local_enforcer->update_tunnel_ids(
+      session_map,
+      create_update_tunnel_ids_request(IMSI1, BEARER_ID_1, teids1));
+  insert_static_rule(1, "", "rule1");
+  RuleRecordTable table;
+  auto record_list = table.mutable_records();
+  // create three rule records version 1, version 2, version 3
+  create_rule_record(
+      IMSI1, test_cfg_.common_context.ue_ipv4(), "rule1", 1, 15, 30, 0, 0,
```
Comment (koolzz) [inline lte/gateway/c/session_manager/test/test_local_enforcer.cpp]: good stuff, can you add some drop bytes as well just to verify that too(the last 2 zeroes in create_rule_record call)?

File: lte/gateway/c/session_manager/test/test_local_enforcer.cpp [hunk 7]
```diff
@@ -429,28 +429,216 @@ TEST_F(LocalEnforcerTest, test_multi_version_reporting) {
...
+  EXPECT_EQ(
+      session_map[IMSI1][0]->get_policy_stats("rule1").current_version, 3);
+  // No calculations used here, stats map stores rx and tx for each version
+  // The older version rule records that are created are disregarded
+  EXPECT_EQ(the_stats.at(1).rx, 20);
+  EXPECT_EQ(the_stats.at(1).tx, 45);
+  EXPECT_EQ(the_stats.at(2).rx, 10);
+  EXPECT_EQ(the_stats.at(2).tx, 80);
+  EXPECT_EQ(the_stats.at(3).rx, 40);
+  EXPECT_EQ(the_stats.at(3).tx, 100);
```
Comment (koolzz) [inline lte/gateway/c/session_manager/test/test_local_enforcer.cpp:517]: Good stuff could you also expect_eq the current/last_reported version?

File: lte/gateway/c/session_manager/test/test_local_enforcer.cpp [hunk 8]
```diff
@@ -429,28 +429,216 @@ TEST_F(LocalEnforcerTest, test_multi_version_reporting) {
...
+  session_map[IMSI1][0]->increment_rule_stats("rule1", &uc);
+  local_enforcer->aggregate_records(session_map, table, update);
+  assert_charging_credit(session_map, IMSI1, SESSION_ID_1, USED_RX, {{1, 70}});
+  assert_charging_credit(session_map, IMSI1, SESSION_ID_1, USED_TX, {{1, 225}});
+  EXPECT_EQ(
+      session_map[IMSI1][0]->get_policy_stats("rule1").current_version, 3);
+}
+
+TEST_F(LocalEnforcerTest, test_erroneous_data) {
+  /*
```
Comment (koolzz) [inline lte/gateway/c/session_manager/test/test_local_enforcer.cpp]: same with comment style

**Review by koolzz — APPROVED** [review]
Added some comments but looks great overall

File: lte/gateway/c/session_manager/test/test_local_enforcer.cpp (hunk 5 above)
Comment (themarwhal) [reply lte/gateway/c/session_manager/test/test_local_enforcer.cpp]: Hey @veshkemburu sorry to nitpick but do you mind using `//` for the comments in test files? (Should prob document this somewhere)

It's often really useful to be able to comment out chunks of tests when debugging so I prefer to have it all be not blocked.

```
// Tests reporting of tx and rx values of old versions, after
// updating to a new version, as well as whether rule records of old
// versions are disregarded appropriately
```

**Comment by themarwhal** [comment]
One last comment about the comment style. The rest looks good to me.

**Review by themarwhal — APPROVED** [review]
lgtm!

**Pull request merged**